### PR TITLE
[1/2]: Overhaul of  donations module

### PR DESCRIPTION
**Pull request by sashko9807**
- Create a new table _payments_, which will hold information to payment itself such as extPaymentIntentId, chargedAmount, amount, billingName, billingEmail etc..
- Refactored _donations_ table to hold only information regarding the donation itself, such as an amount of payment that was allocated to specific campaign, whether that donation is anonymous, whether there is a wish to that donation.
- Created two functions in Vault module, which would allow to increment/decrement many vaults in a single query.
- Adapted existing API endpoints regarding donations, to the new database schema
- Adapted tests to the new database schema

## Motivation and context
Currently two major features are queued:
- Donation to a whole category
- Importing donations coming from Benevity

It was discussed among the PM, and dev team, that our current structure is not fit for situations, where funds from one payment can be allocated to more than one campaign. 

For this reason a new structure was proposed, which makes the relation between a payment and campaign vaults M:N, with _donations_ table being the "middle-man", connection a payment to campaign vault. Figure below"
donation-overall-structure

To make this happen, the current _donations_ structure, was split into two tables _payments_ and _donations_, with relation between those tables being 1:N(one payment, can be linked to one or more donations)
new-donation-structure-FINAL

## Testing
1. Tested whether the current donation records, are safely migrated to the new structure.
2. Tested payment done via Stripe.
3. Tested payment through affiliateCode
4. Tested payment through IRIS import

### Steps to test
1. Before migration, save the current donation records(Open prisma studio->Donations -> Select one row  initially-> Click on the square to select all rows then CTRL+C) in a file.
2. Run the migration script
3. Payments and Donations count should be the exact same
4. Compare the new payment table, with records saved in step1(e.g. match of extPaymentintentId, billingEmail, chargedAmount etc..)
5. Compare the new donations table, with the records saved in step1(e.g. donation type, amount, targetvaultId, personId etc..)
6. Check whether donation via stripe is successfully imported, in the new structure
7. Check whether donation via IRIS bank import is successfully imported in the new structure.
8. Check whether donation via affiliateCode, is successfully imported in the new structure.

### New endpoints
*/donations/payments* - Get information just for the made payments

### CODE REVIEW

File: apps/api/src/donations/donations.service.ts [hunk 1]
```diff
@@ -619,31 +680,36 @@ export class DonationsService {
    * @param updatePaymentDto
    * @returns
    */
-  async update(id: string, updatePaymentDto: UpdatePaymentDto): Promise<Donation> {
+  async update(id: string, updatePaymentDto: UpdatePaymentDto): Promise<Payment> {
     try {
       // execute the below in prisma transaction
       return await this.prisma.$transaction(async (tx) => {
-        const currentDonation = await tx.donation.findFirst({
+        const currentDonation = await tx.payment.findFirst({
```
Comment (igoychev) [inline apps/api/src/donations/donations.service.ts:687]: the variable here should be currentPayment

File: apps/api/src/vault/vault.service.spec.ts [hunk 2]
```diff
@@ -23,4 +28,49 @@ describe('VaultService', () => {
   it('should be defined', () => {
     expect(service).toBeDefined()
   })
+
+  it('decrementManyVaults should throw an error if one vaults returns negative amount', async () => {
```
Comment (igoychev) [inline apps/api/src/vault/vault.service.spec.ts:32]: great addition!

**Review by igoychev — COMMENTED** [review]
all looks great to me and only a few minor comments here&there

File: migrations/20240212112752_new_donation_structure/migration.sql [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+        dbrow.created_at,
+        dbrow.updated_at,
+        dbrow."chargedAmount",
+		dbrow.amount,
+        dbrow.billing_email,
+        dbrow.billing_name
+      ) RETURNING id
+  )
+
+    SELECT * INTO payment_result FROM payment;
```
[the remark this answers is not in the document]
Comment (sashko9807) [reply migrations/20240212112752_new_donation_structure/migration.sql:110]: It's used in the statement below, when inserting the id of the inserted payment row, into `donations` table.

File: apps/api/src/campaign/campaign.controller.spec.ts [hunk 4]
```diff
@@ -428,7 +412,7 @@ describe('CampaignController', () => {
   describe('subscribeToCampaignNotifications', () => {
     it('should throw if no consent is provided', async () => {
       const data: CampaignSubscribeDto = {
-        email: person.email,
+        email: person.email!,
```
[the remark this answers is not in the document]
Comment (sashko9807) [reply apps/api/src/campaign/campaign.controller.spec.ts]: Fixed.

File: apps/api/src/donations/donations.service.ts (hunk 1 above)
Comment (sashko9807) [reply apps/api/src/donations/donations.service.ts]: Done.

File: apps/api/src/bank-transactions/bank-transactions.controller.ts [hunk 5]
```diff
@@ -169,7 +169,8 @@ export class BankTransactionsController {
     const isDev = appEnv === 'development' || appEnv === 'staging'
     if (!isDev) throw new ForbiddenException('Endpoint available only for testing enviroments')
 
-    if (!isAdmin(user)) throw new ForbiddenException('Must be either an admin or active affiliate')
+    if (appEnv === 'staging' && !isAdmin(user))
+      throw new ForbiddenException('Must be either an admin or active affiliate')
```
[the remark this answers is not in the document]
Comment (sashko9807) [reply apps/api/src/bank-transactions/bank-transactions.controller.ts]: Technically the condition above should throw an error on prod environment, but still reverted it to the original logic. 
The initial idea was to make testing easier, on the local environment, but it didn't work as I still needed to attach the token to authorization header.

**Review by igoychev — APPROVED** [review]

**Pull request merged**